packages/flutter/lib/src/widgets/framework.dart

```diff
@@ -1163,17 +1185,205 @@ class _InactiveElements {
+/// Some static functions (e.g. [showDialog], [Theme.of], and so forth) also
+/// take build contexts so that they can act on behalf of the calling widget, or
+/// obtain data specifically for the given context.
+///
+/// Each widget has its own [BuildContext]. In particular, this means that
+/// within a build method, the build context of the widget of the build method
+/// is not the same as the build context of the widgets returned by that build
+/// method. This can lead to some tricky cases. For example, [Theme.of(context)]
+/// looks for the nearest enclosing [Theme] of the given build context. If a
+/// build method for a widget Q includes a [Theme] within its returned widget
+/// tree, and attempts to use [Theme.of] passing its own context, the build
+/// method for Q will not find that [Theme] object. It will instead find
+/// whatever [Theme] was an ancestor to the widget Q. If the build context for a
+/// subpart of the returned tree is needed, a [Builder] widget can be used: the
+/// build context passed to the [Builder.builder] callback will be that of the
+/// [Builder] itself.
+///
+/// For example, in the following snippet, the [ScaffoldState.showSnackBar]
+/// method is called on the [Scaffold] widget that the build method itself
+/// creates. If a [Builder] had not been used, and instead the `context`
+/// argument of the build method itself had been used, no [Scaffold] would have
+/// been found, and the [Scaffold.of] function would have returned null.
+///
+/// ```dart
+///   @override
+///   Widget build(BuildContext context) {
+///     // here, Scaffold.of(context) returns null
+///     return new Scaffold(
+///       appBar: new AppBar(title: new Text('Demo')),
+///       body: new Builder(
+///         builder: (BuildContext context) {
+///           return new FlatButton(
+///             child: new Text('BUTTON'),
+///             onPressed: () {
+///               // here, Scaffold.of(context) returns the locally created Scaffold
+///               Scaffold.of(context).showSnackBar(new SnackBar(
+///                 content: new Text('Hello.')
+///               ));
+///             }
+///           );
+///         }
+///       )
+///     );
+///   }
+/// ```
+///
+/// The [BuildContext] for a particular widget can change location over time as
+/// the widget is moved around the tree. Because of this, values returned from
+/// the methods on this class should not be cached beyond the execution of a
+/// single synchronous function.
 abstract class BuildContext {
+  /// The current configuration of the widget to which this [BuildContext] belongs.
   Widget get widget;
+
+  /// The current [RenderObject] for the widget. If the widget is a
+  /// [RenderObjectWidget], this is the render object that the widget created
+  /// for itself. Otherwise, it is the render object of the first descendant
+  /// [RenderObjectWidget].
+  ///
+  /// This method will only return a valid result after the build phase is
```

Comment:
The references to phases that follow are a little confusing since we haven't explained what all the phases are, how they're orchestrated, how a widget figures out what phase it's in, etc.